lib/spack/spack/package.py

```diff
@@ -2162,6 +2172,68 @@ def do_uninstall(self, force=False):
         # delegate to instance-less method.
         Package.uninstall_by_spec(self.spec, force)
 
+    def do_deprecate(self, replacement, link_fn):
+        """Deprecate this package in favor of replacement spec"""
+        # Copy metadata to deprecated directory of replacement
+        spec = self.spec
+
+        # Check whether package to deprecate has active extensions
+        if self.extendable:
+            view = spack.filesystem_view.YamlFilesystemView(spec.prefix,
+                                                            spack.store.layout)
+            active_exts = view.extensions_layout.extension_map(spec).values()
+            if active_exts:
+                short = spec.format('{name}/{hash:7}')
+                m = "Spec %s has active extensions\n" % short
+                for active in active_exts:
+                    m += '        %s\n' % active.format('{name}/{hash:7}')
+                    m += "Deactivate extensions before deprecating %s" % short
+                tty.die(m)
+
+        # Check whether package to deprecate is an active extension
+        if self.is_extension:
+            tty.msg('AAAAAAAAAA')
+            extendee = self.extendee_spec
+            view = spack.filesystem_view.YamlFilesystemView(extendee.prefix,
+                                                            spack.store.layout)
+
+            if self.is_activated(view):
+                tty.msg('BBBBBBBB')
+                short = spec.format('{name}/{hash:7}')
+                short_ext = extendee.format('{name}/{hash:7}')
+                msg = "Spec %s is an active extension of %s\n" % (short,
+                                                                  short_ext)
+                msg += "Deactivate %s to be able to deprecate it" % short
+                tty.die(msg)
+
+        # Install replacement if it isn't installed already
+        if not spack.store.db.query(replacement):
+            replacement.package.do_install()
+
+        old_deprecator = spack.store.db.deprecator(spec)
+        if old_deprecator:
+            # Find this specs yaml file from its old deprecation
+            self_yaml = spack.store.layout.deprecated_file_path(spec,
+                                                                old_deprecator)
+        else:
+            self_yaml = spack.store.layout.spec_file_path(spec)
+
+        # copy spec metadata to "deprecated" dir of replacement
+        depr_yaml = spack.store.layout.deprecated_file_path(spec,
+                                                            replacement)
+        fs.mkdirp(spack.store.layout.deprecated_dir_path(replacement))
```

Comment:
Can you make this function private and just invoke `dirname` on the output of `deprecated_file_path`? That avoids users having to be aware of multiple functions (although when it comes to calculating the file path it's fine to split it in the logic).